core/app/models/schedule.rb

```diff
@@ -18,10 +19,46 @@ def self.from_sections sections
     end
   end
 
+  def average_start
+      @average_start = 0
+
+      first_periods = Hash.new(9999999999999)
+
+      num_days = 0
+      sections.each do |this_section|
+        this_section.periods.each do |p|
+          num_days += 1.0 if (last_periods[p["day"]] == -1)
+          first_periods[p["day"]] = p["start"].to_i if (p["start"].to_i < first_periods[p["day"]])
+        end
+      end
+      first_periods.each { |day, time| @average_start += time }
+
+      @average_start /= num_days
+  end
+
+  def average_finish
+    @average_finish = 0
```

Comment:
there is an important optimization you can make here. These methods will be called a bunch of times for the sorting, so if the value `@average_finish` already exists, we can simply return it rather than performing the calculation again

Reply:
So do like a
`return @average_finish if @average_finish != 0`